hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinNestedLoopPhysicalRel.java

```diff
@@ -61,9 +58,17 @@ public List<RexNode> rightProjection() {
         return ((HazelcastPhysicalScan) getRight()).projection();
     }
 
-    public JetJoinInfo joinInfo(QueryParameterMetadata parameterMetadata) {
-        List<Integer> leftKeys = analyzeCondition().leftKeys.toIntegerList();
-        List<Integer> rightKeys = analyzeCondition().rightKeys.toIntegerList();
+    @Override
+    public JoinInfo analyzeCondition() {
+        if (modifiedJoinInfo != null) {
+            return modifiedJoinInfo;
+        }
+        JoinInfo joinInfo = super.analyzeCondition();
+        if (getRight().getTable() == null) {
```

Comment:
I didn't get it -- when is it possible?

Reply:
`explainTerms()` is called multiple times during optimization in order to compute hash code for `RelDigest` to be used as a key in `VolcanoPlanner#mapDigestToRel`.  As optimization progresses, join operands become `RelSubset`, `HepRelVertex` and finally some `TableScan`, which causes the first two readings of `getRight().getTable()` to be null. This problem can also be avoided by not adding the condition type in those cases, but it will require the following change, which depends on Calcite internals and may break in the future.
```java
@Override
public RelWriter explainTerms(RelWriter pw) {
    if (getLeft() instanceof RelSubset || getLeft() instanceof HepRelVertex
            || getRight() instanceof RelSubset || getRight() instanceof HepRelVertex) {
        return super.explainTerms(pw);
    }
    return super.explainTerms(pw).item("conditionType",
            !analyzeCondition().rightKeys.isEmpty() ? "equiJoin" : "nonEquiJoin");
}
```